design/one-pager-composition-revisions.md

```diff
@@ -0,0 +1,308 @@
+            region: us-central1
+            settings:
+              tier: db-custom-1-3840
+              dataDiskType: PD_SSD
+      patches:
+        - fromFieldPath: "spec.parameters.storageGB"
+          toFieldPath: "spec.forProvider.settings.dataDiskSizeGb"
+```
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: CompositionRevision
+metadata:
+  # The revision's name is derived by appending the revision number to the
+  # Composition's name.
+  name: example-1
+  labels:
+    crossplane.io/composition-name: example
+    crossplane.io/composition-spec-hash: 707e5bf63687fce7
+  ownerReferences:
+  # The Composition is the controller reference of the revision. When a
+  # Composition is deleted all of its CompositionRevisions are garbage
+  # collected.
+  - apiVersion: apiextensions.crossplane.io/v1
+    kind: Composition
+    name: example
+    controller: true
+spec:
+  # Each revision includes an integer revision number. This number increases
+  # monotonically as the associated composition is updated.
+  revision: 1
+  # Apart from the revision number the rest of the revision spec is identical
+  # to the composition spec.
+  compositeTypeRef:
+    apiVersion: database.example.org/v1alpha1
+    kind: CompositePostgreSQLInstance
+  writeConnectionSecretsToNamespace: crossplane-system
+  resources:
+    - name: cloudsqlinstance
+      base:
+        apiVersion: database.gcp.crossplane.io/v1beta1
+        kind: CloudSQLInstance
+        spec:
+          forProvider:
+            databaseVersion: POSTGRES_9_6
+            region: us-central1
+            settings:
+              tier: db-custom-1-3840
+              dataDiskType: PD_SSD
+      patches:
+        - fromFieldPath: "spec.parameters.storageGB"
+          toFieldPath: "spec.forProvider.settings.dataDiskSizeGb"
+```
+
+Two additional fields will be added to each type of XR and XRC.
+
+```yaml
+apiVersion: database.example.org/v1alpha1
+kind: PostgreSQLInstance
+metadata:
+  name: my-db
+  namespace: default
+spec:
+  parameters:
+    storageGB: 20
+  compositionSelector:
+    matchLabels:
+      provider: gcp
+  compositionRef:
+    name: example-gcp
+  # A new optional field, typically automatically late initialized to the latest
+  # (highest) revision of the referenced composition when the XR.
+  compositionRevisionRef:
+    name: example-gcp-fk2ks
+  # A new optional field. Determines whether the above compositionRevisionRef is
+  # updated automatically to reflect the latest available revision, or whether
+  # it must be updated manually. Valid values are Automatic or Manual. Defaults
+  # to Automatic.
+  compositionUpdatePolicy: Automatic
+```
+
+This design is backward compatible with Crossplane's current behaviour. An XR or
+XRC author may be ignorant of the new `CompositionRevision` type. The XR will
+default to `compositionUpdatePolicy: Automatic`, and thus always select the
+latest revision of the desired `Composition`. Only when the XR or XRC author
```

Comment:
will `Automatic` include any safe guards, or is that (currently) out of scope? For example,
* rate limiting the number of upgrades (this could be difficult without configuration - what is an acceptable number of upgrades per minute for a cluster with 30 XRCs, and one with 10 000?)
* requiring an upgrade to be ready before upgrading the next component (again, I guess this would be difficult, when do we know that the upgrade is done? and for the 10 000 XRC cluster, you probably want upgrades in parallell)

Or will safe guards be `Manual` an external component/process? One helpful thing for an external upgrade process would be to have a status field with the number of ready/unready/synced claims, but I guess this might be difficult to implement?

Reply:
I wasn't planning on building any new safeguards into the `Automatic` strategy. The idea (at least at first) is for `Automatic` to function similarly to the current implementation where there is an implicit "automatic" update of all XRs when a Composition is updated.

That said, the logic that updates the Composition used by each XR will be be built into the XR reconciler. Each reconciler has its own reconcile requeue rate limit, and iirc we have a global rate limit as well. So I'd expect `Automatic` updates to enjoy some safeguards that they did not have previously under this design.